File: quick_sort.java
import java.io.*;
import java.util.*;

public class quick_sort {

    public static void main(String[] args) throws IOException {
        // Check if input filename is provided
        if (args.length != 1) {
            System.out.println("Usage: java quick_sort <dataset_filename.csv>");
            return;
        }

        String filename = args[0];
        List<String[]> tempList = new ArrayList<>();

        // Step 1: Read the dataset from CSV file
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",");
                // Store each row as [integer, string] in a temporary list
                tempList.add(new String[] { parts[0].trim(), parts[1].trim() });
            }
        }

        // Step 2: Convert list to array for better memory handling
        String[][] data = tempList.toArray(new String[0][]);
        tempList.clear(); // Clear the list to free memory

        // Step 3: Sort using Quick Sort (last element as pivot)
        long start = System.nanoTime();
        quickSort(data, 0, data.length - 1); // Perform sorting
        long end = System.nanoTime();

        double timeInSeconds = (end - start) / 1_000_000_000.0;
        System.out.printf("Sorted %d rows in %.4f seconds\n", data.length, timeInSeconds);        

        // Step 4: Write sorted output to new CSV file
        String outFile = "quick_sort_" + data.length + ".csv";
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(outFile))) {
            for (String[] row : data) {
                bw.write(row[0] + "," + row[1]);
                bw.newLine();
            }
        }

        // Step 5: Print execution time
        System.out.println("Sorted " + data.length + " rows in " + (end - start) + " ms");
    }

    // Recursive Quick Sort function using last element as pivot
    public static void quickSort(String[][] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high); // Partition the array
            quickSort(arr, low, pi - 1); // Recursively sort left subarray
            quickSort(arr, pi + 1, high); // Recursively sort right subarray
        }
    }

    // Partition function: places pivot at correct position and rearranges other
    // elements
    public static int partition(String[][] arr, int low, int high) {
        int pivot = Integer.parseInt(arr[high][0]); // Last element as pivot
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (Integer.parseInt(arr[j][0]) < pivot) {
                i++;
                // Swap arr[i] and arr[j]
                String[] temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Place pivot in the correct position
        String[] temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1; // Return the pivot index
    }
}

File: quick_sort.py
import csv
import sys
import time

# Function to read the entire dataset from the CSV file
# Each row is stored as a tuple: (integer, string)
def load_dataset(filename):
    data = []
    with open(filename, 'r') as file:
        reader = csv.reader(file)
        for row in reader:
            data.append((int(row[0]), row[1]))  # Convert first column to int
    return data

# Recursive Quick Sort using last element as pivot
def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)     # Partition the array
        quick_sort(arr, low, pi - 1)       # Sort left half
        quick_sort(arr, pi + 1, high)      # Sort right half

# Partition function for Quick Sort
def partition(arr, low, high):
    pivot = arr[high][0]  # Use integer as pivot
    i = low - 1
    for j in range(low, high):
        if arr[j][0] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]  # Swap smaller element to the left
    arr[i + 1], arr[high] = arr[high], arr[i + 1]  # Place pivot in correct position
    return i + 1

# Write sorted data to output file
def save_sorted_csv(data, filename):
    with open(filename, 'w', newline='') as file:
        writer = csv.writer(file)
        for row in data:
            writer.writerow(row)

def main():
    if len(sys.argv) != 2:
        print("Usage: python quick_sort.py <dataset_filename.csv>")
        return

    input_file = sys.argv[1]

    # Load dataset into memory
    data = load_dataset(input_file)

    # Start timing (in nanoseconds, converted to seconds)
    start_time = time.perf_counter_ns()
    quick_sort(data, 0, len(data) - 1)  # Sort the dataset
    end_time = time.perf_counter_ns()

    # Save sorted output to new CSV file
    output_file = f"quick_sort_{len(data)}.csv"
    save_sorted_csv(data, output_file)

    # Calculate time taken in seconds
    duration_seconds = (end_time - start_time) / 1_000_000_000
    print(f"Sorted {len(data)} rows in {duration_seconds:.4f} seconds")

# Entry point
if __name__ == "__main__":
    main()


